common/ferrostar/src/navigation_controller/mod.rs

```diff
@@ -336,6 +325,31 @@ impl NavigationController {
 
         (current_step_geometry_index, snapped_with_course)
     }
+
+    /// Determines if the navigation controller should advance to the next waypoint.
+    fn should_advance_waypoint(&self, state: &TripState) -> bool {
+        match state {
+            TripState::Navigating {
+                snapped_user_location,
+                ref remaining_waypoints,
+                ..
+            } => {
+                // Update remaining waypoints
+                if let Some(waypoint) = remaining_waypoints.first() {
+                    let current_location: Point = snapped_user_location.coordinates.into();
+                    let next_waypoint: Point = waypoint.coordinate.into();
+                    match self.config.waypoint_advance {
+                        WaypointAdvanceMode::WaypointWithinRange(range) => {
+                            Haversine::distance(current_location, next_waypoint) < range
+                        }
+                    }
+                } else {
+                    false
+                }
```

Comment:
Nice work! I think you can make this even more compact with the [`is_some_and`](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some_and) like this:

```rust
remaining_waypoints.first().is_some_and(|waypoint| {
    // Rest of the inner code
}
```

Then you can drop the whole if/else wrapper and it's a single expression inside :)